compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor.hs

```diff
@@ -98,22 +108,22 @@ checkImports x =
 checkDataTypes :: GHC.ParsedSource -> [(GHC.SrcSpan, String)]
 checkDataTypes m = checkAmbiguousDataTypes m ++ checkUnlabelledConArgs m ++ checkThetas m
 
-
 checkAmbiguousDataTypes :: GHC.ParsedSource -> [(GHC.SrcSpan, String)]
 checkAmbiguousDataTypes (GHC.L _ m) =
     mapMaybe getAmbiguousError (GHC.hsmodDecls m)
   where
     getAmbiguousError :: GHC.LHsDecl GHC.GhcPs -> Maybe (GHC.SrcSpan, String)
     -- Generate an error if the declaration is a data type with one constructor and zero arguments
     getAmbiguousError (GHC.L ss decl)
-      | GHC.TyClD _ GHC.DataDecl{tcdDataDefn=GHC.HsDataDefn{dd_cons=[con]}} <- decl -- single con data type
-      , GHC.PrefixCon [] <- GHC.con_args (GHC.unLoc con) -- zero arguments
+      | GHC.TyClD _ GHC.DataDecl{tcdTyVars = dtyvars, tcdDataDefn=GHC.HsDataDefn{dd_cons=[con]}} <- decl -- single con data type
```

Comment:
We happily accept `data Foo a = Bar | Baz` and turn it into a variant type. However, we don't accept `data Foo a = Bar`. We should fix this inconsistency. I consider the former counterintuitive and would be ok to ban it. To be more precise, I would ban all `data T a_1 ... a_m = C_1 | ... | C_n` for `m >=1` and `n >= 1`. Thoughts, @associahedron @bame-da @cocreature?

Reply:
Once we've decided on this and implemented it, we also need to adjust https://github.com/digital-asset/daml/blob/master/docs/source/app-dev/daml-lf-translation.rst#banned-declarations. (I'm happy to do that myself.)

Reply:
I’d rather lift the restriction in LF so we can have `data Foo a = Bar`. I don’t see why enums can’t have type parameters.

Reply:
The problem is--as always--codegen for **Java**: you cannot have `enum Foo<T> { Bar, Baz }`.

Reply:
:cry: Is there anything stopping us from desugaring `data Foo a = Bar` as a variant? I find phantom type parameters to be quite useful in general so I’d rather not restrict their usage.

Reply:
I find translating `data Foo = Bar` into an enum but `data Foo a = Bar` into a variant quite confusing. If you really want the variant with the phantom type parameter, you can always do `data Foo a = Foo ()`. We should probably stick that into the error message.

Reply:
Fair point, I guess in that case I’m fine with forbidding it but we should probably first turn it into a warning for a couple of releases.

Reply:
I agree with going through the deprecation warning cycle. Ideally, we could capture data on how many users actually hit this warning via our telemetry mechanism. @rohanjr do you have an idea how to do this?

Reply:
The error message already says `Ambiguous data type declaration. Enums cannot have type arguments. Write data Foo a = Bar {} for a record or data Foo a = Bar () for a variant.`

Reply:
I suppose we could re-enable logging for compiler errors/warnings in our telemetry. We used to do that unintentionally and then disabled it, so it shouldn't be too hard to reenable. It may be overkill for this though?

Reply:
We never deliberately logged all compiler errors. We logged some of them by accident (mainly as a result of failures to get scenario results which is not a bug in our code). I think it makes more sense to specifically log certain cases when we introduce the warning rather than logging all errors.

Reply:
It's not a new error though?

`data Foo a = Bar` was not supported before, and continues to be unsupported. The only thing this PR allows that was previously not allowed is `data Foo = Bar`, with no type variables.

I could see us supporting `data Foo a = Bar` in the future, possibly by making the Java codegen smarter about enums.

Reply:
My understanding was that we are talking about introducing an error for `data Foo a = Bar | Baz` which is not an error right now.

Reply:
Ah, I see. Sorry.